packages/expression-parser/src/expression-parser/ExpressionParser.js

```diff
@@ -0,0 +1,110 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
+ */
+
+import { create, all } from 'mathjs';
+
+export class ExpressionParser {
+  constructor() {
+    this.math = create(all, {});
+    this.parser = this.math.parser();
+  }
+
+  /**
+   * Evaluate an expression.
+   * Before evaluating an expression, the scope will be sanitized to fix up any invalid variables that we want to support
+   * @param {*} expression
+   */
+  evaluate(expression) {
+    const originalScope = this.parser.getAll();
+
+    // Sanitize scope and expression to fix any invalid variables before evaluating the expression
+    const sanitizedExpression = this.sanitizeExpression(expression, originalScope);
+
+    // Evaluate the sanitized expression
+    const result = this.parser.evaluate(sanitizedExpression);
+
+    // Reset the scope variables to original;
+    this.clearScope();
+    this.setScope(originalScope);
+
+    return result;
+  }
+
+  /**
+   * Set the parser's scope.
+   * @param {*} scope
+   * @param {*} defaultScope
+   */
+  setScope(scope = {}, defaultScope = {}) {
+    Object.entries(scope).forEach(([name, value]) => {
+      const expressionValue = value || defaultScope[name] || 0;
+      this.set(name, expressionValue);
+    });
+  }
+
+  /**
+   * Add a value to the parser's scope.
+   * @param {*} name
+   * @param {*} value
+   */
+  set(name, value) {
+    this.parser.set(name, value);
+  }
+
+  /**
+   * Sanitize any invalid variables in the scope and expression.
+   * Eg: variables that start with numbers
+   * @param {*} expression
+   * @param {*} originalScope
+   */
+  sanitizeExpression(expression, originalScope) {
```

Comment:
One limitation of mathjs package is it does not allow you to have a variable name starting with `numbers` in an expression (https://github.com/josdejong/mathjs/issues/688#issuecomment-236675020). This is because mathjs supports implicit multiplication (eg: if we have expression `5abc > 3`, it will be parsed as `5 * abc > 3`). This is the default behaviour of mathjs parser and there's no way to turn this off. I tried forking the package, looking at the parser logic and it doesn't seem easy to do this too

In Tupaia I think we do have some cases when we want variables to start with numbers, examples:
- In survey question config, we always want reference questions using uuid so that we can get the answers from the state.
- Some question codes do start with numbers. And we probably always want to use question codes as variables in indicators/report package.

So I was torn with 2 solutions:
1. In any places that use this package, we just have to make sure when constructing expressions, we don't use any variables starting with numbers (eg: in the config of survey question, always add a prefix `$`, formula will look like '$5fa0cdcd5da1e614f30001fd+ $5fa0cdcd5da1e623c30001eg...')

2. Have this package automatically sanitize the variables and expression before actual evaluation.

2nd option seems like the best solution. The downside of 2nd option is we will not be able to implement a util that can get the variables from an expression (similar to `getVariables` in @arithmetic package, which is very handy). Mathjs does support a way of extracting variables from just an expression. But this is only accurate if we don't have any variables starting with numbers in the expression.

So the trade off for the 2nd option is when we construct an expression, we will always have to define a set of variable names expected in the expression. (Eg: in meditrak app, if we use @arithmetic package, we can just grab the variables from the formula and get the values of those variables. Whereas if we use this package, beside the formula, we will also need to store in the config something like:
```
config: {
	formula: '$5fa0cdcd5da1e614f30001fd + $5fa0cdcd5da1e623c30001eg',
	expectedVariables: ['$5fa0cdcd5da1e614f30001fd', '$5fa0cdcd5da1e623c30001eg'],
}
```

What do you reckon?

Reply:
A couple of questions:
- @rohan-bes because of the way these are sanitized, in the data processing pipeline within a report builder, would you have to remember to use `$.row.$5fa0` rather than `$.row.5fa0` in any steps afterwards, or does the context of the expression only last for the single transform step, then we get the result and never see the sanitized variable names again?
- could we implement a `getVariables` which gets the variables from mathjs and then does a desanitize step on the variable names?
- I wonder if we should put a `$` in front of all of the variables in the `scope`, not just those with a leading number? Could cause less confusion if it's a hard rule that we can always expect

Reply:
@edmofro I can answer the last 2 points:

> could we implement a getVariables which gets the variables from mathjs and then does a desanitize step on the variable names?

If what you meant is similar to `getVariables(formula)` in @beyondessential/arithmetic package (which just extracts the variables from the formula without any other info) then yes it's possible. But we have to make sure that the variables in the formula do not start with numbers (Eg: `getVariables('$5fa0 + $4fa1')` -> returns [`$5fa0`, `$4fa1`]). If `getVariables('5fa0 + 4fa1')`, it will return [`fa0`, `fa1`].

> I wonder if we should put a $ in front of all of the variables in the scope, not just those with a leading number? Could cause less confusion if it's a hard rule that we can always expect

This sounds better to me too!

Reply:
Yeah I'm suggesting that we:
- start with `tupaiaExpressionParserInstance.getVariables('5fa0 + 4fa1')`
- sanitize the expression, using your existing logic, creating `'$5fa0 + $4fa1'`
- run mathjs' get variables function, which returns `['$5fa0', '$4fa1']`
- desanitize, stripping the leading `$` -> `['5fa0', '4fa1']`
- return that

Would that work?

Reply:
> sanitize the expression, using your existing logic, creating '$5fa0 + $4fa1'

Yep that would work, but only if the scope already has `5fa0` and `4fa1` variables set in it. This is because the existing sanitize expression logic relies on the variables set in the scope to sanitize the expression (otherwise I don't think there's a way to know what are the variables expected in the formula, unless we use regex which can be complicated since mathjs has a lot of built in syntax)

Reply:
I see what you mean 😞

Reply:
> @rohan-bes because of the way these are sanitized, in the data processing pipeline within a report builder, would you have to remember to use $.row.$5fa0 rather than $.row.5fa0 in any steps afterwards, or does the context of the expression only last for the single transform step, then we get the result and never see the sanitized variable names again?

I believe that `$.row.5fa0` should be fine. 
1. As you say, the expression context is refreshed between each transform step.
2. Even if it weren't, when an object is added to the expression scope (as we do with the `$` object currently in the reporting package) it seems that you are able to reference the children fine even if their names start with a number. ie. `$.row.5fa0` works fine without any extra sanitization

>In any places that use this package, we just have to make sure that when constructing expressions, we don't use any variables starting with numbers (eg: in the config of survey question, always add a prefix $, formula will look like '$5fa0cdcd5da1e614f30001fd+ $5fa0cdcd5da1e623c30001eg...'). This seems very inconvenient and ugly...

@chris-bes I don't necessarily think this is too ugly... Though I don't like the idea of us being forced into using such a syntax simply because of this mathjs implicit multiplication logic...

Reply:
I agree on both points. @rohan-bes and @chris-bes I think this is the major point to resolve before this PR can be approved - could you two work on it together?

Reply:
@edmofro actually I believe I just lied accidentally, I double checked this and `$.row.5fa0` does actually fail unfortunately 😞 You can use `$.row['5fa0']` fine, however that's not intuitive at all...

@chris-bes let's organise some time today to work out a solution for this?

Reply:
Had a discussion with Rohan about this. We agree that:
- The most ideal way to solve this is to modify the parser logic in mathjs to disable the implicit multiplication feature so that we can freely use variables starting with numbers. We would never want to use implicit multiplication in our expressions as well.
- The sanitizing expression approach is probably not the best way and can introduce complications in the future.

Given there's time constraint because both the reporting package and CalculatedQuestion need this soon. I think for now we can:
- Go with the 1st proposed approach above which is just prefix variables with `$` when constructing expressions and remove the sanitizing logic. Also, in this `expression-parser` package, we can add a validate method to detect if there's any implicit multiplication found in the expression and throw an error so that the users know.
- Create a technical debt ticket to fork the mathjs package and disable the implicit multiplication feature? (I'm happy to volunteer to work on it if we agree with this approach).

What do you reckon @edmofro

Reply:
Yeah great conclusions. Thanks guys.